rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,428 @@
+3. The installation mechanism requires 4-5 commands, includes manually installing multiple packages, and varies depending on your version and edition of Teleport.
+4. The use of bash to implement the updater makes changes difficult and prone to error.
+5. The existing auto-updater has limited automated testing.
+6. The use of GPG keys in system package managers has key management implications that we would prefer to solve with TUF in the future.
+7. The desired agent version cannot be set via Teleport's operator-targeted CLI (tctl).
+8. The rollout plan for the new agent version is not fully-configurable using tctl.
+9. Agent installation logic is spread between the auto-updater script, install script, auto-discovery script, and documentation.
+10. Teleport contains logic that is specific to Teleport Cloud upgrade workflows.
+11. The existing auto-updater is not self-updating.
+12. It is difficult and undocumented to automate agent upgrades with custom automation (e.g., with JamF). 
+
+We must provide a seamless, hands-off experience for auto-updates that is easy to maintain.
+
+## Details
+
+We will ship a new auto-updater package written in Go that does not interface with the system package manager.
+It will be distributed as a separate package from Teleport, and manage the installation of the correct Teleport agent version manually. 
+It will read the unauthenticated `/v1/webapi/find` endpoint from the Teleport proxy, parse new fields on that endpoint, and install the specified agent version according to the specified upgrade plan.
+It will download the correct version of Teleport as a tarball, unpack it in `/var/lib/teleport`, and ensure it is symlinked from `/usr/local/bin`.
+
+Source code for the updater will live in `integrations/updater`.
+
+### Installation
+
+```shell
+$ apt-get install teleport-ent-updater
+$ teleport-update enable --proxy example.teleport.sh
+
+# if not enabled already, configure teleport and:
+$ systemctl enable teleport
+```
+
+### API
+
+#### Endpoints
+
+`/v1/webapi/find`
+```json
+{
+  "server_edition": "enterprise",
+  "agent_version": "15.1.1",
+  "agent_auto_update": true,
+  "agent_update_jitter_seconds": 10
+}
+```
+Notes:
+- The Teleport proxy translates upgrade hours (below) into a specific time after which the served `agent_version` changes, resulting in all agents being upgraded.
+- Critical updates are achieved by serving the desired `agent_version` immediately.
+- If an agent misses an upgrade window, it will always update immediately due to the new agent version being served.
+- The edition served is the cluster edition (enterprise, enterprise-fips, or oss), and cannot be configured.
+
+#### Teleport Resources
+
+```yaml
+kind: cluster_maintenance_config
+spec:
+  # agent_auto_update allows turning agent updates on or off at the
+  # cluster level. Only turn agent automatic updates off if self-managed
+  # agent updates are in place.
+  agent_auto_update: true|false
```

Comment:
There is missing days of week in the schedule that are critical to most customers.

Reply:
There has to be some way to partition updates based on labels, so not all agents will get updated at the same time. This is critical at scale.

I would imagine:

```yaml
agent_auto_update:
   enabled: true
   hours: 0-23
   days_of_week: [sat, sun]
   strategy:
     rolling_upgrade:
         window: 10% # update only 10% at a time
         labels:  
            stage:
               hours: 0-23
               days_of_week: [*]
            prod:
               hours: 5-6
               days_of_week: [sat, sun]
```

Reply:
Will add, was just chatting with @russjones about these requirements.

Reply:
Jotting down initial requirements from customer. Will discuss further.

* Control day of week
* Control by label
* Approval required before upgrade begins rolling out

Reply:
Do you have more info on this requirement? Is it enough to allow the customer to enable/disable auto-upgrades and create their own schedule, or do we need, e.g., an "approve upgrade" button that appears in the UI before the window?